### DISCUSSION TITLE
Adding year label "a" to the first of the duplicates?

### DISCUSSION THREAD

**Discussion by ryofurue** [retorquere/zotero-better-bibtex#1916] [Q&A] [answered]
I use this format
````
[auth.etal+initials:capitalize]:[year]
````
for citation keys.  The problem is that when there is a name clash, with two citations ending up in `AuthorA.etal:2017`, the second one becomes `AuthorA.etal:2017a`.

This is confusing to me when I cite these papers in my document.  For the reference list, most journals in my field list these two as
````
Author A, . . .  2017a . . . 
Author A, . . .  2017b . . .
````
so that the key `AuthorA.etal:2017a` corresponds to the "2017b" paper!

I wish the extra year label "a" were added to the first citekey.  If that's hard, I wish the extra label started from "b".

How do you achieve that?

**Comment by retorquere** [comment, marked as answer]
`[postfix+1=%(a)s]`

**Comment by ryofurue** [comment]
@retorquere  Thank you!
````
[auth.etal+initials:capitalize]:[year][postfix+1=%(a)s]
````
gives cite keys I wanted.  (I have no idea how this "language" works.  Perhaps `postfix` is an integer counter which starts from 0. `+1` increments it. `=` means "convert to" and `%(a)` is like LaTeX's `\alph{postfix}`, but what's `s` at the end? . . . I'm just curious. Nothing more than that.)

**Reply by retorquere** [reply]
`postfix` indeed starts from `0`, where `0` means "don't use one". `+1` is actually a fixed value, `+2` or anything else doesn't work, but that's the idea behind it, yes. the alpha postfix is the "excel column name of column `n`", so if you go past 26 you get `aa`, `ab`, etc.

The `s` comes from sprintf-js; from the BBT docs (no criticism, I just need to provide context for the explanation that follows):

> postfix=<spec>/postfix+1=<spec>: a pseudo-function that sets the citekey disambiguation postfix using an [sprintf-js](https://www.npmjs.com/package/sprintf-js) format spec for when a key is generated that already exists. Does not add any text to the citekey otherwise. You must include exactly one of the placeholders %(n)s (number), %(a)s (alpha, lowercase) or %(A)s (alpha, uppercase). For the rest of the disambiguator you can use things like padding and extra text as sprintf-js allows. With +1 the disambiguator is always included, even if there is no need for it because no duplicates exist. The default format is %(a)s. 

I opted for sprintf-js because these days I try not to invent a new language from scratch for everything; `%(a)s` means "return the sprintf-variable `a`, which is a *s*tring". `a` has the generated postfix (the letter). There is also the string-variable `A`, which has upper-cased `a`, and the variable `n`, which is numeric. The names of the variables are chosen by me, and are sort of arbitrary, they could have been anything, but `a` for postfixes `a`, `b`, etc and `n` for `1`, `2`, etc made sense to me. They're all strings, because the number is also returned as a string, but you could for example do 

* `%(a)'a3s` to have 3-position alpha postfixes like `aad`, or
* `%(n)03s` to have 3-position numeric postfixes like `008`.

using sprintf you can then add whatever fixed text you want `:%(a)s` to get `:b` etc.

**Reply by ryofurue** [reply]
@retorquere  Thank you very much for the detailed explanation and sorry for not having looked at the documentation.  I found the documentation site extremely complete and looking sharp!

Now I've found that my setting
````
[auth.etal+initials:capitalize]:[year][postfix+1=%(a)s]
````
is wrong!  because most citekeys now acquire "a" at the end even though there aren't duplicates!

What I originally wanted was the following.

If there are no duplicates:
````
SmithM:2020
````
If there are duplicates:
````
SmithM:2020a
SmithM:2020b
. . .
````
But, perhaps this scheme cannot be implemented, at least using this "language".

Having said that, I have the feeling that my scheme is fundamentally flawed.  If you find another M. Smith paper that comes between 2020a and 2020b, you have to change existing citekeys and your existing LaTeX documents break.  Perhaps I should give up on the idea of clean neat suffixes and learn to live with the default behavior of BBT.
 
(It's still early days for me with Zotero, and so I can afford to change citekeys.)

**Reply by retorquere** [reply]
> sorry for not having looked at the documentation. I found the documentation site extremely complete and looking sharp!

That's great to hear but no apologies are necessary. BBT has grown complex, and you have to know how the documentation is structured to find these things. I organize for clarity where I can, but I do not expect people to have studied these docs in depth.

>  If you find another M. Smith paper that comes between 2020a and 2020b, you have to change existing citekeys and your existing LaTeX documents break.

This is why this behavior isn't in BBT. I have had one user that anticipated a lot of these clashes, and always wanted keys to have these postfixes. I don't use this myself.

A major objective of BBT is to keep your keys stable, so there is only one -- configurable -- instance where BBT will change your key when the item doesn't change, and that is when you have:

* An item with a dynamic key `smith2021`
* You later assign a pinned key `smith2021` to another item

In which case BBT needs you to have made a choice on how to handle this. Do you want:

1. the keys to be stable, accepting that both now have `smith2021` for a key, or
2. unique keys, accepting that `smith2021` is changed to `smith2021a` in the background, even though that item did not change

This is only when the original `smith2021` is not pinned. If you pin a key, I don't second-guess your choice, and you can have duplicates all you want.

The default behavior of BBT is scenario 1.
